File: binaryTrees/35.py
from typing import List


class TreeNode:
    def __init__(self, val: int):
        self.val = val
        self.left = None
        self.right = None


def solution(root: TreeNode, cont: List[int]):
    if root is None:
        return (0, 0)

    sol_left = solution(root.left, cont)
    sol_right = solution(root.right, cont)

    sum_nodes = sol_left[0] + sol_right[0] + root.val
    cant_nodes = sol_left[1] + sol_right[1] + 1

    if sum_nodes // cant_nodes == root.val:
        cont[0] += 1

    return (sum_nodes, cant_nodes)


def averageOfSubtree(root: TreeNode) -> int:
    cont = [0]

    solution(root, cont)

    return cont[0]


if __name__ == "__main__":
    root = TreeNode(4)
    root.left = TreeNode(8)
    root.right = TreeNode(5)
    root.left.left = TreeNode(0)
    root.left.right = TreeNode(1)
    root.right.right = TreeNode(6)

    print(averageOfSubtree(root))

File: binaryTrees/35.cpp
#include "../common/tree.h"

using namespace Tree;

pair<int, int> solution(TreeNode *root, int &cont) {
    if (root ==nullptr) {
        return make_pair(0, 0);
    }

    auto solLeft = solution(root->left, cont);
    auto solRight = solution(root->right, cont);

    int cantNodes = solLeft.second + solRight.second + 1;
    int sumNodes = solLeft.first + solRight.first + root->val;

    if(sumNodes / cantNodes == root->val) cont +=1;

    return make_pair(sumNodes, cantNodes);
}


int averageOfSubtree(TreeNode* root) {
    int cont = 0;

    solution(root, cont);

    return cont;
        
}

int main() {
    TreeNode *root = createNode(1);
    root->left = createNode(8);
    root->right = createNode(5);
    root->left->left = createNode(0);
    root->left->right = createNode(1);
    root->right->right = createNode(6);

    cout << averageOfSubtree(root);

    return 0;
}